File: src/agentrust_telemetry/evidence.py
"""Deterministic, pre-sampling evidence accumulation."""

from __future__ import annotations

import hashlib
import json
from copy import deepcopy
from dataclasses import dataclass
from threading import Lock
from typing import Any, Callable, Literal

from .errors import EvidenceError, EvidencePersistenceError
from .validation import SchemaValidator


Completeness = Literal["complete", "incomplete", "unknown"]
CANONICALIZATION_PROFILE = "agentrust-json-v1"
_GENESIS_DIGEST = bytes(32)


@dataclass(frozen=True)
class EvidenceEntry:
    sequence: int
    event_id: str
    previous_digest: str | None
    digest: str
    event: dict[str, Any]


@dataclass(frozen=True)
class EvidenceSnapshot:
    run_id: str
    entries: tuple[EvidenceEntry, ...]
    chain_digest: str | None
    canonicalization_profile: str
    completeness: Completeness
    sealed: bool


DurableAppend = Callable[[EvidenceEntry], bool]


class EvidenceAccumulator:
    """Accumulate one run's validated events before operational projection.

    A durable callback must be idempotent by ``event_id`` and return ``True`` only
    after the entry is durably committed. Callback failure leaves local state
    unchanged so callers can retry the same event.
    """

    def __init__(
        self,
        run_id: str,
        validator: SchemaValidator,
        *,
        durable_append: DurableAppend | None = None,
        max_events: int = 10_000,
    ) -> None:
        if not isinstance(run_id, str) or not run_id:
            raise EvidenceError("run_id must be a non-empty string")
        if max_events < 1:
            raise EvidenceError("max_events must be at least 1")
        self._run_id = run_id
        self._validator = validator
        self._durable_append = durable_append
        self._max_events = max_events
        self._entries: list[EvidenceEntry] = []
        self._event_ids: set[str] = set()
        self._sealed = False
        self._completeness: Completeness = "unknown"
        self._lock = Lock()

    @property
    def mode(self) -> Literal["memory", "callback"]:
        return "callback" if self._durable_append is not None else "memory"

    def append(self, event: dict[str, Any]) -> EvidenceEntry:
        """Validate, chain, durably acknowledge, then accept an event."""
        self._validator.validate(event)
        event_copy = deepcopy(event)
        if event_copy["run_id"] != self._run_id:
            raise EvidenceError("event run_id does not match accumulator run_id")

        with self._lock:
            if self._sealed:
                raise EvidenceError("evidence run is sealed")
            event_id = event_copy["event_id"]
            if event_id in self._event_ids:
                raise EvidenceError(f"duplicate event_id: {event_id}")
            if len(self._entries) >= self._max_events:
                raise EvidenceError(f"evidence run exceeds max_events={self._max_events}")

            sequence = len(self._entries)
            previous = self._entries[-1].digest if self._entries else None
            try:
                digest = _entry_digest(sequence, previous, event_copy)
            except (TypeError, ValueError) as exc:
                raise EvidenceError(
                    f"event_id={event_id} cannot be canonicalized under "
                    f"{CANONICALIZATION_PROFILE}"
                ) from exc
            entry = EvidenceEntry(sequence, event_id, previous, digest, event_copy)

            if self._durable_append is not None:
                try:
                    acknowledged = self._durable_append(_copy_entry(entry))
                except Exception as exc:
                    raise EvidencePersistenceError(
                        f"durable evidence callback failed for event_id={event_id}"
                    ) from exc
                if acknowledged is not True:
                    raise EvidencePersistenceError(
                        f"durable evidence callback did not acknowledge event_id={event_id}"
                    )

            self._entries.append(_copy_entry(entry))
            self._event_ids.add(event_id)
            return _copy_entry(entry)

    def seal(self, *, completeness: Completeness) -> EvidenceSnapshot:
        """Close the run with a caller-asserted completeness assessment."""
        if completeness not in ("complete", "incomplete", "unknown"):
            raise EvidenceError(f"unsupported completeness: {completeness!r}")
        with self._lock:
            if self._sealed:
                raise EvidenceError("evidence run is already sealed")
            self._sealed = True
            self._completeness = completeness
            return self._snapshot()

    def snapshot(self) -> EvidenceSnapshot:
        with self._lock:
            return self._snapshot()

    def _snapshot(self) -> EvidenceSnapshot:
        entries = tuple(_copy_entry(item) for item in self._entries)
        return EvidenceSnapshot(
            run_id=self._run_id,
            entries=entries,
            chain_digest=entries[-1].digest if entries else None,
            canonicalization_profile=CANONICALIZATION_PROFILE,
            completeness=self._completeness if self._sealed else "unknown",
            sealed=self._sealed,
        )


def _entry_digest(sequence: int, previous_digest: str | None, event: dict[str, Any]) -> str:
    canonical = json.dumps(
        event,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    ).encode("utf-8")
    previous = bytes.fromhex(previous_digest) if previous_digest else _GENESIS_DIGEST
    material = previous + sequence.to_bytes(8, "big") + canonical
    return hashlib.sha256(material).hexdigest()


def _copy_entry(entry: EvidenceEntry) -> EvidenceEntry:
    return EvidenceEntry(
        entry.sequence,
        entry.event_id,
        entry.previous_digest,
        entry.digest,
        deepcopy(entry.event),
    )

File: packages/typescript/src/evidence.ts
import {createHash} from "node:crypto";
import canonicalize from "canonicalize";
import type {NormalizedEvent} from "./types.js";
import {SchemaValidator} from "./validation.js";

export type Completeness = "complete" | "incomplete" | "unknown";
export const CANONICALIZATION_PROFILE = "rfc8785-jcs-v1";
const GENESIS_DIGEST = Buffer.alloc(32);

export class EvidenceError extends Error {}
export class EvidencePersistenceError extends EvidenceError {}

export interface EvidenceEntry {
  sequence: number;
  eventId: string;
  previousDigest?: string;
  digest: string;
  event: NormalizedEvent;
}

export interface EvidenceSnapshot {
  runId: string;
  entries: readonly EvidenceEntry[];
  chainDigest?: string;
  canonicalizationProfile: typeof CANONICALIZATION_PROFILE;
  completeness: Completeness;
  sealed: boolean;
}

export type DurableAppend = (entry: EvidenceEntry) => boolean;

export class EvidenceAccumulator {
  readonly #entries: EvidenceEntry[] = [];
  readonly #eventIds = new Set<string>();
  #sealed = false;
  #completeness: Completeness = "unknown";
  #mutating = false;

  constructor(
    readonly runId: string,
    readonly validator: SchemaValidator,
    readonly options: {durableAppend?: DurableAppend; maxEvents?: number} = {},
  ) {
    if (!runId) throw new EvidenceError("runId must be a non-empty string");
    if ((options.maxEvents ?? 10_000) < 1) throw new EvidenceError("maxEvents must be at least 1");
  }

  get mode(): "memory" | "callback" { return this.options.durableAppend ? "callback" : "memory"; }

  append(event: NormalizedEvent): EvidenceEntry {
    this.validator.validate(event);
    const eventCopy = structuredClone(event);
    if (eventCopy.run_id !== this.runId) throw new EvidenceError("event run_id does not match accumulator runId");
    if (this.#mutating) throw new EvidenceError("reentrant evidence mutation is not supported");
    this.#mutating = true;
    try {
      if (this.#sealed) throw new EvidenceError("evidence run is sealed");
      if (this.#eventIds.has(eventCopy.event_id)) throw new EvidenceError(`duplicate event_id: ${eventCopy.event_id}`);
      if (this.#entries.length >= (this.options.maxEvents ?? 10_000)) throw new EvidenceError(`evidence run exceeds maxEvents=${this.options.maxEvents ?? 10_000}`);
      const sequence = this.#entries.length;
      const previousDigest = this.#entries.at(-1)?.digest;
      let digest: string;
      try { digest = entryDigest(sequence, previousDigest, eventCopy); }
      catch (error) { throw new EvidenceError(`event_id=${eventCopy.event_id} cannot be canonicalized under ${CANONICALIZATION_PROFILE}`, {cause: error}); }
      const entry: EvidenceEntry = {sequence, eventId: eventCopy.event_id, ...(previousDigest ? {previousDigest} : {}), digest, event: eventCopy};
      if (this.options.durableAppend) {
        let acknowledged: boolean;
        try { acknowledged = this.options.durableAppend(cloneEntry(entry)); }
        catch (error) { throw new EvidencePersistenceError(`durable evidence callback failed for event_id=${entry.eventId}`, {cause: error}); }
        if (acknowledged !== true) throw new EvidencePersistenceError(`durable evidence callback did not acknowledge event_id=${entry.eventId}`);
      }
      this.#entries.push(cloneEntry(entry));
      this.#eventIds.add(entry.eventId);
      return cloneEntry(entry);
    } finally { this.#mutating = false; }
  }

  seal(completeness: Completeness): EvidenceSnapshot {
    if (!(["complete", "incomplete", "unknown"] as const).includes(completeness)) throw new EvidenceError(`unsupported completeness: ${String(completeness)}`);
    if (this.#mutating) throw new EvidenceError("reentrant evidence mutation is not supported");
    if (this.#sealed) throw new EvidenceError("evidence run is already sealed");
    this.#sealed = true;
    this.#completeness = completeness;
    return this.snapshot();
  }

  snapshot(): EvidenceSnapshot {
    const entries = this.#entries.map(cloneEntry);
    const chainDigest = entries.at(-1)?.digest;
    return {runId: this.runId, entries, ...(chainDigest ? {chainDigest} : {}), canonicalizationProfile: CANONICALIZATION_PROFILE, completeness: this.#sealed ? this.#completeness : "unknown", sealed: this.#sealed};
  }
}

function entryDigest(sequence: number, previousDigest: string | undefined, event: NormalizedEvent): string {
  const canonical = canonicalize(event);
  if (canonical === undefined) throw new TypeError("event has no canonical JSON representation");
  const sequenceBytes = Buffer.alloc(8);
  sequenceBytes.writeBigUInt64BE(BigInt(sequence));
  const previous = previousDigest ? Buffer.from(previousDigest, "hex") : GENESIS_DIGEST;
  if (previous.length !== 32) throw new TypeError("previous digest must be 32 bytes");
  return createHash("sha256").update(previous).update(sequenceBytes).update(canonical, "utf8").digest("hex");
}

function cloneEntry(entry: EvidenceEntry): EvidenceEntry { return structuredClone(entry); }
